### ISSUE TITLE
Performance regression from #48044

### ISSUE THREAD

**Issue by epmatsw** [microsoft/TypeScript#52345]
# Bug Report

It seems like a fairly significant performance regression was introduced in this https://github.com/microsoft/TypeScript/pull/48044, which was shipped in TS 4.7. In TS 4.6, it takes ~30 seconds to run `yarn tsc` in our repo, and the reported strict subtype cache size is 24_011. In TS 4.7, it takes 75+ seconds to run `yarn tsc` (~2.5x longer than previous), and the reported strict subtype cache size is 122_778 (~5x larger than previous).

I verified that it was that specific PR that regressed the performance by taking a locally-installed 4.6.4 and applying the changes from that PR directly.

Another symptom is that several types (mostly spreads of objects into others) are now reporting as "Expression produces a union type that is too complex to represent."

### 🔎 Search Terms

`Expression produces a union type that is too complex to represent`, `strict subtype cache size`

### 🕗 Version & Regression Information

- Regressed in https://github.com/microsoft/TypeScript/pull/48044
- This changed between versions 4.6.4 and 4.7.3

|Version|Time|Strict Subtype Cache Size|
|-------|-----|--------------------------|
|4.6.4  | 33s | 24011                                  |
|4.7.3  | 75s | 122778                                 |
|4.9.4  | 77s | 122869                                |
|5.0.0-dev.20230120 | 78s | 122869 |

### 💻 Code

I haven't been able to narrow down a minimal reproduction of this issue, but if someone points me in the right direction I'm happy to work on it.

### 🙁 Actual behavior

Running `yarn tsc --noEmit --extendedDiagnostics` takes ~75 seconds in our repo, with a Strict subtype cache size of 122778.

### 🙂 Expected behavior

Running `yarn tsc --noEmit --extendedDiagnostics` takes ~30 seconds in our repo, with a Strict subtype cache size of 24011.

**Comment by DanielRosenwasser** [comment]
Would you be able to use our analyze-trace tool?

https://github.com/microsoft/typescript-analyze-trace

It should highlight specific hot spots that might occur as of TS 4.7. The [usage instructions](https://github.com/microsoft/typescript-analyze-trace#usage) are a little better now. Let me know if you need any more info using it.

**Comment by epmatsw** [comment]
Not public, unfortunately. Sorry about the delay on the trace, I’ll try to get something going tomorrow.

**Comment by epmatsw** [comment]
Dang, thought I had a reproduction but apparently not. Probably not gonna get something until Monday now :(

**Comment by epmatsw** [comment]
Okay, so I'm not sure this is minimal, but I think it's on the right track:

```typescript
type R = `${number}a` & {
    _thing: true;
};

type _S = "1" | "2" | "3" | "4" | "5";
type S = `${_S}${_S}`;

// Produces "Expression produces a union type that is too complex to represent"
// type S = `${_S}${_S}${_S}`;


type T = R | S;
type X = `${T} ${T}`;

export type Props = Partial<{
    x: X;
}>;

const a1: Props = {};
const a2: Props = {};

const b = { ...a1, ...a2 };

export { b };
```

In 4.6.4, that has a Strict subtype cache size of 0. In 4.7.4, that has a Strict subtype cache size of 34476.

**Comment by epmatsw** [comment]
Hmmm. I guess because X boils down to `string` in 4.6.4. So I'm not sure how useful that even actually is other than "actually checking this gigantic type rather than not checking it is expensive" :/

**Comment by jakebailey** [comment]
That might be enough, actually. By modifiying it a bit to try and make it worse (but not enough to make it go to `string`), I can get the test to take a good 5-6 seconds on my machine, wheras in v4.6, it takes under a second.

```ts
type R = `${number}a` & {
    _thing: true;
};

type _S = "1" | "2" | "3" | "4" | "5" | "6";

type S = `${_S}${_S}${_S}`;


type T = R | S;
type X = `${T} ${T}`;

export type Props = Partial<{
    x: X;
}>;

const a1: Props = {};
const a2: Props = {};

const b = { ...a1, ...a2 };

export { b };
```

**Comment by jakebailey** [comment]
After staring at this for a while, I'm at a loss for a way to speed this up; we're now actually traversing the intersections whereas before, we weren't. This example in particular turns into a combinatorial explosion after #48044, which I didn't really see coming at the time.

I'm not sure what the path forward is at the moment, besides reverting that PR (or doing nothing), but I'll keep trying.

**Comment by jakebailey** [comment]
@epmatsw Would you mind testing the package on my PR for this issue here? https://github.com/microsoft/TypeScript/pull/52836#issuecomment-1435492877

I believe it fixes the performance problem (in that it avoids massive intersections), but, it does make these sorts of constructs stricter (more correct in some situations) and the above sample in particular will still give an error that complains about it being too big.

**Comment by epmatsw** [comment]
Unfortunately not a huge improvement:

TS 4.6.4
```
Strict subtype cache size:    24040
Total time:                  34.40s
```

TS Version 5.0.0-insiders.20230218 (https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/146938/artifacts?artifactName=tgz&fileId=E6600AFB90EEF24611686B292C01E0D427A0C8AA2B90E3E4B7E85880B3CF614102&fileName=/typescript-5.0.0-insiders.20230218.tgz):
```
Strict subtype cache size:   122673
Total time:                  65.03s
```

**Comment by jakebailey** [comment]
Just to confirm, it's ~13s faster than a previous 5.0 build, correct?

**Comment by jakebailey** [comment]
But, good to know that it's not a complete fix; I'll need to retry with the test case and check the subtype cache metric, as you had said it was previously zero, so that's next to look at.

**Comment by epmatsw** [comment]
Unfortunately I think that's my bad (I had a fairly significant hardware upgrade since the first set of measurements).

Testing on the 20230120 build from before, I get ~64s now, so no significant difference from the 20230218 build.

**Comment by jakebailey** [comment]
Hm, that's unfortunate; it means that the test case I've been using is not representative of your actual code.

Have you been able to use the analyze-trace tool to try and pinpoint the exact expression that is problematic?

Also, we are able sign NDAs (or reuse existing ones) in order to look at private code, if that's palatable. @DanielRosenwasser knows that process better than me, I think.

**Comment by epmatsw** [comment]
Sorry for the radio silence for a bit, sick kid from daycare unfortunately. I'm hoping I'll get some time to tinker on this more this week.

**Comment by epmatsw** [comment]
Well, the only thing I could really glean from the trace is that it's only places where we spread 2 object of the type together. So in the example from https://github.com/microsoft/TypeScript/issues/52345#issuecomment-1407216732, where `Props` is the expensive type, the flagged code looks like:

```ts
const Component = ({ p }: { p?: Props }) => {
	const p2: Props = {};
	return <Component2 p={{ ...p, ...p2 }} />;
}
```

Always 2+ spreads of the Props type.

Also confirmed that 5.0.1rc, 5.1.0-dev.20230303, and the packed build from the PR all have roughly identical performance :(

**Comment by jakebailey** [comment]
I've merged #52836, which you said unfortunately didn't help; the PR was updated a little bit since you tested so I'll be interested whether or not tonight's nightly build will improve things even a little.

**Comment by jakebailey** [comment]
@epmatsw Another one for you to try, when you get a chance: https://github.com/microsoft/TypeScript/pull/53406

The package should show up in a few minutes.

**Comment by epmatsw** [comment]
Oooh nice one.

4.6.4 was 28.9+-1.2
5.0.2 wsa 64.4+-0.6
that preview was 47.6+-0.4

**Comment by epmatsw** [comment]
I can run #53413 if you pack it too

**Comment by epmatsw** [comment]
Also a nice one:

4.6.4: 26.9s
5.0.2: 64.4s
53406: 47.3s
53413: 48.2s

4.6.4:
Assignability cache size:   1235198
Identity cache size:          24716
Subtype cache size:          103233
Strict subtype cache size:    50192

5.0.2:
Assignability cache size:   1219203
Identity cache size:          37601
Subtype cache size:          135593
Strict subtype cache size:   182779

53406:
Assignability cache size:   1219412
Identity cache size:          37610
Subtype cache size:          135757
Strict subtype cache size:   117180

53413:
Assignability cache size:   1219412
Identity cache size:          37610
Subtype cache size:          135757
Strict subtype cache size:   117180

**Comment by jakebailey** [comment]
Cool, thanks. Theoretically, they should compound, but #53406 isn't quite ready yet.

**Comment by jakebailey** [comment]
One last to test, for now, which is #53406 now that main has #53413, so we can test both together: https://github.com/microsoft/TypeScript/pull/53406#issuecomment-1478631677

**Comment by epmatsw** [comment]
Not much change from the previous two, 47.1s. So a little faster but within noise level.

Assignability cache size:   1219412
Identity cache size:          37610
Subtype cache size:          135757
Strict subtype cache size:   117180

**Comment by jakebailey** [comment]
Example usage would be:

```
$ pprof-it ./node_modules/typescript/lib/tsc.js ...
```

**Comment by epmatsw** [comment]
In the meantime, from profiling more now that some of the noise is gone, it seems like the remaining hotspots boil down to code that looks like:

```ts
type Piece = "A" | "B" | "C" | "D" | "E";
type Key = `${Piece}.${Piece}.${Piece}.${Piece}.${Piece}`;
type GetPrefixes<T> = T extends `${infer U}.A` ? U : never;
type Prefix = GetPrefixes<Key>;
```

**Comment by epmatsw** [comment]
Okay, I finally got back to this (so sorry for the delay!).

I got a really minimal reproduction of the slowdown: https://github.com/epmatsw/tsdemo

There's two workspaces, one with ts4.6 and one with ts5.2. There's two tsconfigs, one that includes a "slow" file and one that includes a "fast" file. The only difference between the slow and fast files is that one has an extra `as any`:

<img width="375" alt="image" src="https://github.com/microsoft/TypeScript/assets/1126580/0cbd6d71-0118-4789-8721-c318fa86e8d9">

That `as any` doesn't make any difference in TS4.6, but in TS5.2 it results in a ~13x speedup (bringing the number back to TS4.6 levels):

<img width="538" alt="image" src="https://github.com/microsoft/TypeScript/assets/1126580/5f73cd91-87b9-4fc9-b8dd-b1d9f659cacd">

**Comment by epmatsw** [comment]
Fwiw, I also found a single `any` that when added to our codebase reduced our `tsc` time from ~70s to ~50s. Presumably there's another one or two instances of this slowdown that accounts for the remaining 20s of regression that I may eventually be able to track down, but it (potentially naively) *seems* like there may be some optimization that could be made here.

**Comment by jakebailey** [comment]
I just got a chance to look at this; this case doesn't look anything like your previous example; it doesn't involve intersections / template literals and can't be caused by the same thing.

The code is fast in both 4.6 and 4.7 (the original issue started happening in 4.7), but gets 10x slower in 4.8, then 2x slower still in 5.0.

It might be worth filing this case separately as it'll involve different stuff. I'll have to bisect it, but it's definitely not the original issue here, which was more "we actually start handling intersections within template expressions instead of converting them to `string` straight away", compared to this example, which is more like a combinatoric explosion from attempting to narrow a union of 10,000 strings against itself.

**Comment by jakebailey** [comment]
https://github.com/microsoft/TypeScript/pull/55926 technically resolves this case; a PR build will be available soon. Though I'm not totally sure how I feel about this particular condition just because it only works when you attempt to narrow something to itself, which feels rare and is only causing a perf hole here because you're narrowing a huge union to itself. If you ever change one of them, this optimization fails again.

**Comment by epmatsw** [comment]
Makes sense! I didn't even think about bisecting it as another issue, I was too excited to finally have tracked it down haha. I can re-file for sure if you want a clean slate to work through this, just let me know.

I think that fix makes sense and definitely doesn't hurt, but I also think it's partly because I over-narrowed my example. I re-expanded it to be more like the actual code, and the performance with the fix is now (as you predicted) back to the slower time. The general pattern looks like:

```ts
function a (someLiteral: BigLiteral | SmallLiteral) {
    if (isBigLiteral(someLiteral)) {
        // do BigLiteral stuff
   } else {
        // do SmallLiteral stuff
   }
}
```

**Comment by jakebailey** [comment]
Do you actually have a union of 10,000 strings in your codebase? Or is this closer to a toy example?

**Comment by epmatsw** [comment]
Unfortunately we do. It’s `keyof typeof <10k line JSON file>`

**Comment by epmatsw** [comment]
Although in that case I think `SmallLiteral` is more like 20-100 literals rather than 1000 (which I just picked for roundness in the example)

**Comment by jakebailey** [comment]
Just because I've been working on a tool to make it easy to bisect any TS version _ever_, I checked and the change that 10x'd this in 4.8 was #49625, and the change that 2x'd on top of that in 5.0 was #53085 (cherry pick of #52984). Neither are surprising.

But yeah, I would class this a different issue to file than the original case (which, I'd still like to investigate).

**Comment by epmatsw** [comment]
Filed in https://github.com/microsoft/TypeScript/issues/55948!

**Comment by jakebailey** [comment]
So barring the stuff in #55948, is there anything left here that isn't fast? I've gone through the code bits in this thread and they have all improved, and I can't really recall what's left here. It'd be nice of course to actually be able to test the real world example... I am of course happy to test it and if privacy is a concern I'm pretty sure we do NDAs.

**Comment by epmatsw** [comment]
I think it's safe to call this one fixed! The remaining regression since 4.6 is fully accounted for by the issue we're tracking in the other thread.
